### ISSUE TITLE
/m modifier in RE causes segmentation fault

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#1940]
Migrated from [rt.perl.org#3210](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=3210) (status was 'resolved')

Searchable as RT3210$

**Comment by p5pRT** [comment]
### From baghdne@electra.chiron.com

#### Created by baghdne@electra.chiron.com



apparently\, a search regexp with the /mg modifiers causes a segmentation
fault on long strings\.
eg&#8203;:

$filecontent= \`cat file\` ;
$names = $filecontents =~ /^\(\\w\+\)\\\(/mg ;

this works fine on perl 5\.005\_03 on the same platform

thanks

nehmeh\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.6.0:

Configured by baghdne at Wed May  3 09:20:36 PDT 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=irix, osvers=6.5, archname=IP27-irix
    uname='irix64 electra 6.5 07151432 ip27 '
    config_args='-de -Dprefix=/products/shared/local -Dinstallusrbinperl=undef'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc -n32', optimize='-O3', gccversion=
    cppflags='-D_BSD_TYPES -D_BSD_TIME -OPT:Olimit=0:space=ON -I/usr/local/include -DLANGUAGE_C'
    ccflags ='-D_BSD_TYPES -D_BSD_TIME -woff 1009,1110,1174,1184,1552 -OPT:Olimit=0:space=ON -I/usr/local/include -DLANGUAGE_C'
    stdchar='unsigned char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc -n32', ldflags =' -L/usr/local/lib32 -L/usr/local/lib -Wl,-woff,84'
    libpth=/usr/local/lib /usr/lib32 /lib32 /lib /usr/lib
    libs=-ldb -lm -lc
    libc=/usr/lib32/libc.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-n32 -shared -L/usr/local/lib32 -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /products/shared/local/lib/perl5/5.6.0/IP27-irix
    /products/shared/local/lib/perl5/5.6.0
    /products/shared/local/lib/perl5/site_perl/5.6.0/IP27-irix
    /products/shared/local/lib/perl5/site_perl/5.6.0
    /products/shared/local/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/user/baghdne
    LANG=C
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/local/package/daylight-461/v461/lib:
    LOGDIR (unset)
    PATH=/products/shared/local/bin:/protos/package/Tripos/trigo:/local/package/grid18:/usr/sbin:/usr/bsd:/sbin:/usr/bin:/usr/bin/X11:.:/usr/local/bin:/usr/freeware/bin:/usr/lib/print:/opt/budTool/bin:/usr/etc:/usr/etc/appletalk:/local/package/daylight-461/v461/bin:/protos/package/moe/bin:/protos/package/cdrprogs/bin:/protos/package/golpe:/protos/package/Tripos/Mix/DiverseSolutions/DVSolns
    PERL_BADLANG (unset)
    SHELL=/usr/bin/csh


```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Nehmeh Baghdadi writes&#8203;:
> apparently\, a search regexp with the /mg modifiers causes a segmentation
> fault on long strings\.

> $filecontent= \`cat file\` ;
> $names = $filecontents =~ /^\(\\w\+\)\\\(/mg ;

How long is long?

~/perl/perl\-5\.6\.0\->\./perl \-wle \\
 '$\_ = \("xx " x 1000000\)\. "yy\(\) "\. \("xx " x 100000\); $names = /^\(\\w\+\)\\\(/mg ;'

shows no problem\.

Ilya

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Hi

thanks for replying\.

the file contents span multiple lines\.             
basically\, this is a \.c files\, and the regexp is supposed to grab function
names\. My files size is around 60\,000 chars\, much less than what you
tried\. I am begining to suspect that this is due to some flaw while building
perl on my platform\.

thanks\.

nehmeh

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Fri\, May 05\, 2000 at 04&#8203;:11&#8203;:58PM \-0700\, Nehmeh Baghdadi wrote&#8203;:
> the file contents span multiple lines\.             
> basically\, this is a \.c files\, and the regexp is supposed to grab function
> names\. My files size is around 60\,000 chars\, much less than what you
> tried\. I am begining to suspect that this is due to some flaw while building
> perl on my platform\.

I would not be surprised if this is a genuine bug\.  But we need a way
to reproduce it\.  I tried

\./perl \-wle '$\_ = \("xx xx\\n" x 10000\)\. "yy\(\) "\. \("xx xx\\n" x 10000\); $names = /^\(\\w\+\)\\\(/mg ;'

but it still finished OK\.  Though very slow\.  I would suspect a
quadratic behaviour due to a misunderstanding of the optimizer\.

Ilya

**Comment by p5pRT** [comment]
### From @vanstyn

In \<20000505193409\.B679@&#8203;monk\.mps\.ohio\-state\.edu>\, Ilya Zakharevich writes&#8203;:
:On Fri\, May 05\, 2000 at 04&#8203;:11&#8203;:58PM \-0700\, Nehmeh Baghdadi wrote&#8203;:
:> the file contents span multiple lines\.             
:> basically\, this is a \.c files\, and the regexp is supposed to grab function
:> names\. My files size is around 60\,000 chars\, much less than what you
:> tried\. I am begining to suspect that this is due to some flaw while building
:> perl on my platform\.
:
:I would not be surprised if this is a genuine bug\.  But we need a way
:to reproduce it\.  I tried
:
:\./perl \-wle '$\_ = \("xx xx\\n" x 10000\)\. "yy\(\) "\. \("xx xx\\n" x 10000\); $names = 
:/^\(\\w\+\)\\\(/mg ;'
:
:but it still finished OK\.  Though very slow\.  I would suspect a
:quadratic behaviour due to a misunderstanding of the optimizer\.

Take lines 2\-167 of toke\.c from perl\-5\.6\.0\, and you can get a
reproducible SEGV&#8203;:

crypt% head \-167 toke\.c | perl \-pwe 'BEGIN\{\<>\} s/\\\(/ /g unless eof' >testfile
crypt% perl \-w
$file = \`cat testfile\`;
$match = $file =~ /^\(\\w\+\)\\\(/mg;
\_\_END\_\_
Name "main&#8203;::match" used only once&#8203;: possible typo at \- line 2\.
Segmentation fault \(core dumped\)
crypt% 

\(Removing all but the last '\(' shortens the path to the SEGV by avoiding
early matches from fbm\_instr\(\)\. More radical changes to the data tend
to suppress the SEGV\. The last line is the first match for the regexp&#8203;:
"S\_ao\(pTHX\_ int toketype\)\\n"\.\)

The SEGV occurs at regexec\.c line 631&#8203;:
&nbsp;      set\_useful&#8203;:
&nbsp;        \+\+BmUSEFUL\(prog\->check\_substr\); /\* hooray/5 \*/
\.\. because prog\->check\_substr is NULL\, because intuit\_start has already
reached 'success\_at\_start&#8203;:' \(line 653\)\, and taken the path that sets&#8203;:
&nbsp;            prog\->check\_substr = Nullsv;        /\* disable \*/

It then proceeds through the following 'Last resort' code \- I don't
understand why\, nor why it declares that 'This position contradicts
STCLASS' \- until it gets sent back towards the main loop by the
'goto try\_at\_offset' \(line 747\) which drops immediately through to
'set\_useful' and the SEGV\.

I have no idea what is going on here\, so I hope this helps\.

Hugo

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

At 23&#8203;:34 \+0100 2000\-05\-07\, Hugo wrote&#8203;:
>Take lines 2\-167 of toke\.c from perl\-5\.6\.0\, and you can get a
>reproducible SEGV&#8203;:
>
>crypt% head \-167 toke\.c | perl \-pwe 'BEGIN\{\<>\} s/\\\(/ /g unless eof' >testfile
>crypt% perl \-w
>$file = \`cat testfile\`;
>$match = \($file =~ /^\(\\w\+\)\\\(/mg\);
>\_\_END\_\_
>Name "main&#8203;::match" used only once&#8203;: possible typo at \- line 2\.
>Segmentation fault \(core dumped\)
>crypt%

FWIW\, does not segfault on MachTen\, but does under LinuxPPC\.  With 
MachTen\, I get

Name "main&#8203;::match" used only once&#8203;: possible typo at \- line 2\.
Use of uninitialized value in pattern match \(m//\) at \- line 2\.

While you scratch your head about the core dump\, I can scratch mine 
about that weird second warning\.  Have not investigated further\, 
however\.

I thought this might be a stack overflow thing\, but no&#8203;: recursion's 
not happening\, is it?  Next possibility's a wild pointer&#8203;: as it's 
bereft of memory protection\, MachTen may let its apps keep going 
under circumstances where less cavalier environments \-\- like Linux \-\- 
would have shot them down\.

>The SEGV occurs at regexec\.c line 631&#8203;:

It's probably occurring there under LinuxPPC too&#8203;: gdb shows that the 
core was produced by a blow\-up somewhere in Perl\_re\_intuit\_start\(\); I 
don't have a debugging perl handy to zero in more closely\.

What's your platform?   Here's a summary of my MachTen perl5 
\(revision 5\.0 version 6 subversion 0\) configuration&#8203;:
&nbsp;   Platform&#8203;:
&nbsp;     osname=machten\, osvers=4\.1\.4\, archname=powerpc\-machten
&nbsp;     uname='machten ppp100 5 0\.5 powerpc '
&nbsp;     config\_args='\-des'
&nbsp;     hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;     usethreads=undef use5005threads=undef useithreads=undef 
usemultiplicity=unde
f
&nbsp;     useperlio=undef d\_sfio=undef uselargefiles=define
&nbsp;     use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
&nbsp;   Compiler&#8203;:
&nbsp;     cc='cc'\, optimize='\-O2 \-fomit\-frame\-pointer'\, gccversion=2\.8\.1
&nbsp;     cppflags='\-DNOTDEF\_MACHTEN \-DREG\_INFTY=2047 \-I/usr/local/include'
&nbsp;     ccflags ='\-DNOTDEF\_MACHTEN \-DREG\_INFTY=2047 \-I/usr/local/include'
&nbsp;     stdchar='char'\, d\_stdstdio=undef\, usevfork=true
&nbsp;     intsize=4\, longsize=4\, ptrsize=4\, doublesize=8
&nbsp;     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=8
&nbsp;     ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, 
Off\_t='off\_t'\, lseeksize
=4
&nbsp;     alignbytes=8\, usemymalloc=y\, prototype=define
&nbsp;   Linker and Libraries&#8203;:
&nbsp;     ld='ld'\, ldflags =' \-Xlstack=1048576 \-L/usr/local/lib'
&nbsp;     libpth=/usr/local/lib /usr/lib
&nbsp;     libs=\-lndbm \-lgdbm \-ldb \-lm \-lc
&nbsp;     libc=/usr/lib/libc\.a\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;   Dynamic Linking&#8203;:
&nbsp;     dlsrc=dl\_none\.xs\, dlext=none\, d\_dlsymun=undef\, ccdlflags=''
&nbsp;     cccdlflags=''\, lddlflags=''

**Comment by p5pRT** [comment]
### From @vanstyn

In \<p04310101b53c17c1e740@&#8203;\[192\.168\.1\.4\]>\, Dominic Dunlop writes&#8203;:
:At 23&#8203;:34 \+0100 2000\-05\-07\, Hugo wrote&#8203;:
:>Take lines 2\-167 of toke\.c from perl\-5\.6\.0\, and you can get a
:>reproducible SEGV&#8203;:
:>
:>crypt% head \-167 toke\.c | perl \-pwe 'BEGIN\{\<>\} s/\\\(/ /g unless eof' >testfile
:>crypt% perl \-w
:>$file = \`cat testfile\`;
:>$match = \($file =~ /^\(\\w\+\)\\\(/mg\);
:>\_\_END\_\_
:>Name "main&#8203;::match" used only once&#8203;: possible typo at \- line 2\.
:>Segmentation fault \(core dumped\)
:>crypt%
:
:FWIW\, does not segfault on MachTen\, but does under LinuxPPC\.  With 
:MachTen\, I get
:
:Name "main&#8203;::match" used only once&#8203;: possible typo at \- line 2\.
:Use of uninitialized value in pattern match \(m//\) at \- line 2\.

I believe the warning is an alternate expression of the same problem\,
possibly due to calling fbm\_instr\(\) after nulling out prog\->check\_substr\.
In my testfile\, I get the undef warning as well if I don't white out
the early parens\, and in other situations I see only the undef warning\.

It is possible that \_no\_ more work should be done after success\_at\_start\,
in which case this patch should fix the problem&#8203;:

<details><summary>Inline Patch</summary>

```diff
--- regexec.c.old	Sun Mar  5 08:53:08 2000
+++ regexec.c	Mon May  8 14:09:43 2000
@@ -667,2 +667,3 @@
 	    prog->reganch &= ~RE_USE_INTUIT;
+	    return s;
 	}


```
</details>
.. but I suspect that will remove some optimisations that re_intuit_start
is supposed to do\, and that a subtler fix is preferred\.

:What's your platform?

Hugo


---
Summary of my perl5 \(revision 5\.0 version 6 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.2\.5\-16\, archname=i686\-linux\-64int
&nbsp;    uname='linux crypt\.compulink\.co\.uk 2\.2\.5\-16 \#1 sun may 30 23&#8203;:00&#8203;:18 bst 1999 i686 unknown '
&nbsp;    config\_args='\-des \-Doptimize=\-g \-O6 \-Dprefix=/opt/perl\-5\.6\.0\.64bit \-Duse64bitint'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=undef d\_sfio=undef uselargefiles=define 
&nbsp;    use64bitint=define use64bitall=undef uselongdouble=undef usesocks=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, optimize='\-g \-O6'\, gccversion=egcs\-2\.91\.66 19990314/Linux \(egcs\-1\.1\.2 release\)
&nbsp;    cppflags='\-DDEBUGGING \-fno\-strict\-aliasing'
&nbsp;    ccflags ='\-DDEBUGGING \-fno\-strict\-aliasing \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'
&nbsp;    stdchar='char'\, d\_stdstdio=define\, usevfork=false
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, usemymalloc=n\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-lndbm \-lgdbm \-ldb \-ldl \-lm \-lc \-lposix \-lcrypt
&nbsp;    libc=/lib/libc\-2\.1\.1\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic'
&nbsp;    cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: DEBUGGING USE\_64\_BIT\_INT USE\_LARGE\_FILES
&nbsp;  Built under linux
&nbsp;  Compiled at Mar 23 2000 11&#8203;:46&#8203;:28
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /opt/perl\-5\.6\.0\.64bit/lib/5\.6\.0/i686\-linux\-64int
&nbsp;    /opt/perl\-5\.6\.0\.64bit/lib/5\.6\.0
&nbsp;    /opt/perl\-5\.6\.0\.64bit/lib/site\_perl/5\.6\.0/i686\-linux\-64int
&nbsp;    /opt/perl\-5\.6\.0\.64bit/lib/site\_perl/5\.6\.0
&nbsp;    /opt/perl\-5\.6\.0\.64bit/lib/site\_perl
&nbsp;    \.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Mon\, May 08\, 2000 at 02&#8203;:20&#8203;:52PM \+0100\, Hugo wrote&#8203;:
> \-\-\- regexec\.c\.old    Sun Mar  5 08&#8203;:53&#8203;:08 2000
> \+\+\+ regexec\.c    Mon May  8 14&#8203;:09&#8203;:43 2000
> @&#8203;@&#8203; \-667\,2 \+667\,3 @&#8203;@&#8203;
>          prog\->reganch &= ~RE\_USE\_INTUIT;
> \+        return s;
>      \}

You just disabled the first\-char\-cognizance optimization in this case\.

Ilya

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Sun\, May 07\, 2000 at 11&#8203;:34&#8203;:42PM \+0100\, Hugo wrote&#8203;:
> It then proceeds through the following 'Last resort' code \- I don't
> understand why\,

Do not understand why \*what\*?  It tries a following hint from the sniffer\.  It
would be wiser to postpone deletion of the string until all the hints
fail to abort search\, but it may be tricky\.

But most probably it is the best solution\.\.\.  Anyone ready to try?

> nor why it declares that 'This position contradicts STCLASS'

Maybe because it does?  ;\-\)  The sniffer found out that the start of match
should be

&nbsp;  a\) before '\)'
&nbsp;  b\) at /^/m
&nbsp;  c\) before \[a\-zA\-Z\_0\-9\]\.

After the 'b' hint is used\, 'c' is checked\.

Ilya

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

>apparently\, a search regexp with the /mg modifiers causes a segmentation
>fault on long strings\.
>eg&#8203;:
>
>$filecontent= \`cat file\` ;
>$names = $filecontents =~ /^\(\\w\+\)\\\(/mg ;

FWIW\, below my \.signature is the Purify run on this with Perl v5\.6\.0 on
HP\-UX 10\.20\.
===============================================
Mark Leighton Fisher            fisherm@&#8203;tce\.com
Thomson Consumer Electronics    Indianapolis IN
"Display some adaptability\." \-\- Doug Shaftoe\, \_Cryptonomicon\_


\*\*\*\*  Purify instrumented pureperl \(pid 24057 at Mon May  8 15&#8203;:11&#8203;:03 2000\)
&nbsp;  \* Purify 5\.0\.1 HP\-UX\, Copyright \(C\) 1992\-1999 Rational Software Corp\. All
rights reserved\. 
&nbsp;  \* For contact information type&#8203;: "purify \-help"
&nbsp;  \* For TTY output\, use the option "\-windows=no"
&nbsp;  \* Command\-line&#8203;: pureperl \-w 
&nbsp;  \* Options settings&#8203;: \-purify \-purify\-home=/usr/pure/purify\-5\.0\.1\-hpux 

\*\*\*\*  Purify instrumented pureperl \(pid 24057\)  \*\*\*\*
NPR&#8203;: Null pointer read&#8203;:
&nbsp;  \* This is occurring while in&#8203;:
&nbsp;    Perl\_re\_intuit\_start \[regexec\.c&#8203;:631\]
&nbsp;    Perl\_regexec\_flags \[regexec\.c&#8203;:1405\]
&nbsp;    Perl\_pp\_match  \[pp\_hot\.c&#8203;:1027\]
&nbsp;    Perl\_runops\_debug \[run\.c&#8203;:54\]
&nbsp;    S\_run\_body     \[perl\.c&#8203;:1401\]
&nbsp;    perl\_run       \[perl\.c&#8203;:1325\]
&nbsp;  \* Reading 4 bytes from 0x0

\*\*\*\*  Purify instrumented pureperl \(pid 24057\)  \*\*\*\*
ZPR&#8203;: Zero page read&#8203;:
&nbsp;  \* This is occurring while in&#8203;:
&nbsp;    Perl\_re\_intuit\_start \[regexec\.c&#8203;:631\]
&nbsp;    Perl\_regexec\_flags \[regexec\.c&#8203;:1405\]
&nbsp;    Perl\_pp\_match  \[pp\_hot\.c&#8203;:1027\]
&nbsp;    Perl\_runops\_debug \[run\.c&#8203;:54\]
&nbsp;    S\_run\_body     \[perl\.c&#8203;:1401\]
&nbsp;    perl\_run       \[perl\.c&#8203;:1325\]
&nbsp;  \* Reading 4 bytes from 0x20

\*\*\*\*  Purify instrumented pureperl \(pid 24057\)  \*\*\*\*
ZPW&#8203;: Zero page write&#8203;:
&nbsp;  \* This is occurring while in&#8203;:
&nbsp;    Perl\_re\_intuit\_start \[regexec\.c&#8203;:631\]
&nbsp;    Perl\_regexec\_flags \[regexec\.c&#8203;:1405\]
&nbsp;    Perl\_pp\_match  \[pp\_hot\.c&#8203;:1027\]
&nbsp;    Perl\_runops\_debug \[run\.c&#8203;:54\]
&nbsp;    S\_run\_body     \[perl\.c&#8203;:1401\]
&nbsp;    perl\_run       \[perl\.c&#8203;:1325\]
&nbsp;  \* Writing 4 bytes to 0x20

\*\*\*\*  Purify instrumented pureperl \(pid 24057\)  \*\*\*\*
COR&#8203;: Fatal core dump&#8203;:
&nbsp;  \* This is occurring while in&#8203;:
&nbsp;    Perl\_re\_intuit\_start \[regexec\.c&#8203;:631\]
&nbsp;    Perl\_regexec\_flags \[regexec\.c&#8203;:1405\]
&nbsp;    Perl\_pp\_match  \[pp\_hot\.c&#8203;:1027\]
&nbsp;    Perl\_runops\_debug \[run\.c&#8203;:54\]
&nbsp;    S\_run\_body     \[perl\.c&#8203;:1401\]
&nbsp;    perl\_run       \[perl\.c&#8203;:1325\]
&nbsp;  \* Received signal 10 \(SIGBUS \- Bus error\)
&nbsp;  \* Handler function&#8203;: SIG\_DFL
&nbsp;  \* Faulting address = 0x20
&nbsp;  \* Signal mask&#8203;: 
&nbsp;  \* Pending signals&#8203;:

**Comment by p5pRT** [comment]
### From @vanstyn

In \<20000508150204\.A12241@&#8203;monk\.mps\.ohio\-state\.edu>\, Ilya Zakharevich writes&#8203;:
:On Sun\, May 07\, 2000 at 11&#8203;:34&#8203;:42PM \+0100\, Hugo wrote&#8203;:
:> It then proceeds through the following 'Last resort' code \- I don't
:> understand why\,
:
:Do not understand why \*what\*?

The phrase 'last resort' normally implies 'everything else we've tried
has failed'; however\, we've just come from a label 'success\_at\_start'\,
so I guess either the comment or the label's name is misleading\.

Even so\, since I understand hardly any of this routine I suspect my
confusion in this small thing is just a symptom of a greater confusion\.

Hugo

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Mon\, May 08\, 2000 at 11&#8203;:22&#8203;:43PM \+0100\, Hugo wrote&#8203;:
> In \<20000508150204\.A12241@&#8203;monk\.mps\.ohio\-state\.edu>\, Ilya Zakharevich writes&#8203;:
> :On Sun\, May 07\, 2000 at 11&#8203;:34&#8203;:42PM \+0100\, Hugo wrote&#8203;:
> :> It then proceeds through the following 'Last resort' code \- I don't
> :> understand why\,
> :
> :Do not understand why \*what\*?
> 
> The phrase 'last resort' normally implies 'everything else we've tried
> has failed'; however\, we've just come from a label 'success\_at\_start'\,
> so I guess either the comment or the label's name is misleading\.

But success of the matcher means a failure of the optimizer\!  ;\-\) The
whole idea of optimizer is quickly to filter out as many
start\-positions as possible\.

Ilya

**Comment by p5pRT** [comment]
### From @vanstyn

In \<20000508215433\.A16176@&#8203;monk\.mps\.ohio\-state\.edu>\, Ilya Zakharevich writes&#8203;:
:On Mon\, May 08\, 2000 at 11&#8203;:22&#8203;:43PM \+0100\, Hugo wrote&#8203;:
:> In \<20000508150204\.A12241@&#8203;monk\.mps\.ohio\-state\.edu>\, Ilya Zakharevich writes&#8203;:
:> :On Sun\, May 07\, 2000 at 11&#8203;:34&#8203;:42PM \+0100\, Hugo wrote&#8203;:
:> :> It then proceeds through the following 'Last resort' code \- I don't
:> :> understand why\,
:> :
:> :Do not understand why \*what\*?
:> 
:> The phrase 'last resort' normally implies 'everything else we've tried
:> has failed'; however\, we've just come from a label 'success\_at\_start'\,
:> so I guess either the comment or the label's name is misleading\.
:
:But success of the matcher means a failure of the optimizer\!  ;\-\) The
:whole idea of optimizer is quickly to filter out as many
:start\-positions as possible\.

I had no idea it was in 'the matcher'\. Let me try to explain better
why I find this routine confusing\, by annotating a chunk of it with
some specific questions and confusions&#8203;: I've attached that below\.

I am not asking you to answer the questions\, or to explain what is
going on; I am\, rather\, trying to highlight why the code is so
difficult to understand\.

I do not know what the solution is&#8203;: perhaps none is needed\, and I am
the only one to find this code unreasonably opaque\. However at a
rough count there are some 180 non\-debugging statements in the
routine \(spread across 480 lines\)\, of which 30 are gotos going to
12 different labels \- for me these are a serious barrier to
understanding the flow\.

When I have time enough\, I'll continue trying to unravel the code;
perhaps I'll get to the point I understand it well enough to offer
patches remedying some of my problems\.

Hugo
\-\-\- regexec\.c&#8203;:339\-400
:    if \(prog\->reganch & ROPT\_ANCH\) \{    /\* Match at beg\-of\-str or after \\n \*/
:    ml\_anch = \!\( \(prog\->reganch & ROPT\_ANCH\_SINGLE\)
:             || \( \(prog\->reganch & ROPT\_ANCH\_BOL\)
:              && \!PL\_multiline \) \);    /\* Check after \\n? \*/

I have no idea why this particular expression means "check after \\n"\.
I guess that for some reason GPOS and SBOL suppress multiline; or
maybe some combinations of these flags can't exist\. 

:    if \(\(prog\->check\_offset\_min == prog\->check\_offset\_max\) && \!ml\_anch\) \{

Why only if \!ml\_anch? Maybe the check\_offsets are relative to the
anchor\, but then I'd expect \!GPOS to be checked as well\.

:        /\* Substring at constant offset from beg\-of\-str\.\.\. \*/
:        I32 slen;
:
:        if \( \!\(prog\->reganch & ROPT\_ANCH\_GPOS\) /\* Checked by the caller \*/
:         && \(sv && \(strpos \+ SvCUR\(sv\) \!= strend\)\) \) \{

After following sv back to regexec\_flags\, and then looking at various
references to CALLREGEXEC \(once I'd remembered its name\)\, I eventually
worked out that sv would contain the string being checked\.

:        DEBUG\_r\(PerlIO\_printf\(Perl\_debug\_log\, "Not at start\.\.\.\\n"\)\);
:        goto fail;
:        \}
:        PL\_regeol = strend;            /\* Used in HOP\(\) \*/
:        s = HOPc\(strpos\, prog\->check\_offset\_min\);
:        if \(SvTAIL\(check\)\) \{

I have no idea what this does&#8203;: there is no explanation to be found where
SvTAIL is defined\, nor where SVpbm\_tail is defined\, nor in the region
where check\_substr is defined\. I guess the next step would be to look
for all other places in the code where SvTAIL is touched\.

:        slen = SvCUR\(check\);    /\* >= 1 \*/

I guess that comment tells you something\.

:        if \( strend \- s > slen || strend \- s \< slen \- 1 
:             || \(strend \- s == slen && strend\[\-1\] \!= '\\n'\)\) \{
:            DEBUG\_r\(PerlIO\_printf\(Perl\_debug\_log\, "String too long\.\.\.\\n"\)\);

'String too long or too short'\, presumably\.

:            goto fail\_finish;

I've seen a label 'fail' and a label 'fail\_finish'\. Apparently the
only difference is that fail\_finish also increases BmUSEFUL\, but
the names don't give any clue to that\.

:        \}
:        /\* Now should match s\[0\.\.slen\-2\] \*/

Why?

:        slen\-\-;
:        if \(slen && \(\*SvPVX\(check\) \!= \*s
:                 || \(slen > 1
:                 && memNE\(SvPVX\(check\)\, s\, slen\)\)\)\) \{

Hmm\, why check the first character separately from the rest of the
string? If it is an optimisation\, it should probably be part of the
macro itself\, or of another macro\.

:          report\_neq&#8203;:
:            DEBUG\_r\(PerlIO\_printf\(Perl\_debug\_log\, "String not equal\.\.\.\\n"\)\);
:            goto fail\_finish;
:        \}
:        \}
:        else if \(\*SvPVX\(check\) \!= \*s
:             || \(\(slen = SvCUR\(check\)\) > 1
:             && memNE\(SvPVX\(check\)\, s\, slen\)\)\)

Again\, \(memNE\(SvPVX\(check\)\, s\, SvCUR\(check\)\) seems much clearer\.

:        goto report\_neq;

This label apparently exists only to avoid duplicating a DEBUG\_r\(\) call\.
That doesn't seem important enough to make the code yet more convoluted\.

:        goto success\_at\_start;
:    \}
:    /\* Match is anchored\, but substr is not anchored wrt beg\-of\-str\. \*/
:    s = strpos;
:    start\_shift = prog\->check\_offset\_min; /\* okay to underestimate on CC \*/

I have no idea what CC is\, which makes this comment unhelpful\. My guess
is that it has something to do with the possibility that multibyte
characters may appear early in the string\.

:    end\_shift = prog\->minlen \- start\_shift \-
:        CHR\_SVLEN\(check\) \+ \(SvTAIL\(check\) \!= 0\);

No idea\. If I could guess what 'end\_shift' was likely to contain\,
I could perhaps attempt to guess what the code intended to do\.

:    if \(\!ml\_anch\) \{
:        I32 end = prog\->check\_offset\_max \+ CHR\_SVLEN\(check\)
:                     \- \(SvTAIL\(check\) \!= 0\);
:        I32 eshift = strend \- s \- end;
:
:        if \(end\_shift \< eshift\)
:        end\_shift = eshift;
:    \}
:    \}
:    else \{                /\* Can match at random position \*/
:    ml\_anch = 0;
:    s = strpos;
:    start\_shift = prog\->check\_offset\_min; /\* okay to underestimate on CC \*/
:    /\* Should be nonnegative\! \*/

\.\. or what? Does this mean 'it is logically guaranteed that the value will
be nonnegative'\, or 'we'll coredump if it ever is'\, or what?

:    end\_shift = prog\->minlen \- start\_shift \-
:        CHR\_SVLEN\(check\) \+ \(SvTAIL\(check\) \!= 0\);
:    \}

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Tue\, May 09\, 2000 at 04&#8203;:39&#8203;:44AM \+0100\, Hugo wrote&#8203;:
> I do not know what the solution is&#8203;: perhaps none is needed\, and I am
> the only one to find this code unreasonably opaque\. However at a
> rough count there are some 180 non\-debugging statements in the
> routine \(spread across 480 lines\)\, of which 30 are gotos going to
> 12 different labels \- for me these are a serious barrier to
> understanding the flow\.

There is some objective need for complicated flow\.  We may apply 4
different checks for each REx&#8203;: anchored substring\, floating substring\,
BOL anchor\, and start\-charclass\.  If one of this fails\, then the
current position cannot be a match \(which is good\!\)\, so we may need to
restart search from the next position\.  But which check should be
applied at the next position?  This depends on which features were
found by the sniffer\, \*and\* on the relative position of the floating
substring and of the current point\.

So the flow cannot be \*much\* less complicated than it is\.  But of
course\, improvements are possible and strongly wanted\.

> When I have time enough\, I'll continue trying to unravel the code;
> perhaps I'll get to the point I understand it well enough to offer
> patches remedying some of my problems\.

Do not consider the current code as anything more than a
proof\-of\-concept\.  It was created in purely incremental way\, by adding
more and more new things to check\.  Additionally\, a lot of choices done
there could \(thus should\) have been done at compile time and encoded
in flags\.

Ilya

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Tue\, May 09\, 2000 at 04&#8203;:39&#8203;:44AM \+0100\, Hugo wrote&#8203;:
> :    if \(prog\->reganch & ROPT\_ANCH\) \{    /\* Match at beg\-of\-str or after \\n \*/
> :    ml\_anch = \!\( \(prog\->reganch & ROPT\_ANCH\_SINGLE\)
> :             || \( \(prog\->reganch & ROPT\_ANCH\_BOL\)
> :              && \!PL\_multiline \) \);    /\* Check after \\n? \*/

ANCH is an OR of several possibilities\, including GPOS=\\G \(which \*can\*
match anywhere ;\-\)\.  But ROPT\_ANCH\_GPOS is already processed by the
caller\, who would provide pos\(\) as the place to start lookup\.

> :     if \(\(prog\->check\_offset\_min == prog\->check\_offset\_max\) && \!ml\_anch\) \{
>
> Why only if \!ml\_anch?
> Maybe the check\_offsets are relative to the
> anchor\, but then I'd expect \!GPOS to be checked as well\.

Offsets are relative to start\-of\-search\.  The caller ensures
start\-of\-search for GPOS\.

> :         if \(SvTAIL\(check\)\) \{
>
> I have no idea what this does

Each sub\-string can have a form 'foo'$\.  SvTAIL is responsible for $
\(currently \\Z and \\z are not special\-cased\)\.

> where check\_substr is defined\. I guess the next step would be to look
> for all other places in the code where SvTAIL is touched\.

Better look on code for debugging output\.  ;\-\)

> :        if \( strend \- s > slen || strend \- s \< slen \- 1 
> :             || \(strend \- s == slen && strend\[\-1\] \!= '\\n'\)\) \{
> :            DEBUG\_r\(PerlIO\_printf\(Perl\_debug\_log\, "String too long\.\.\.\\n"\)\);
> 
> 'String too long or too short'\, presumably\.

With an additional twist that the internal representation of 'foo'$ is
fbm'ed "foo\\n"\, so it has an extra 1 in the length\.

> :        /\* Now should match s\[0\.\.slen\-2\] \*/
> 
> Why?

See above\.

> :        goto report\_neq;
> 
> This label apparently exists only to avoid duplicating a DEBUG\_r\(\) call\.
> That doesn't seem important enough to make the code yet more convoluted\.

You mean that better debugging output "doesn't seem important enough"? ;\-\)

> :    start\_shift = prog\->check\_offset\_min; /\* okay to underestimate on CC \*/
> 
> I have no idea what CC is

Me too\.  And Larry did not explain when I asked\.  I assume that it is
some indication of byte vs char\, but i could not guess \*which\* one\.

> :    end\_shift = prog\->minlen \- start\_shift \-
> :        CHR\_SVLEN\(check\) \+ \(SvTAIL\(check\) \!= 0\);
> 
> No idea\. If I could guess what 'end\_shift' was likely to contain\,
> I could perhaps attempt to guess what the code intended to do\.

How many chars at the end of the string cannot get into fixed substrings\.

> :    start\_shift = prog\->check\_offset\_min; /\* okay to underestimate on CC \*/
> :    /\* Should be nonnegative\! \*/
> 
> \.\. or what?

Or future code is busted\.\.\.

> Does this mean 'it is logically guaranteed that the value will
> be nonnegative'\, or 'we'll coredump if it ever is'\, or what?

Both\, for some value of "logically"\.

Ilya

**Comment by p5pRT** [comment]
### From @TimToady

Ilya Zakharevich writes&#8203;:
: > :    start\_shift = prog\->check\_offset\_min; /\* okay to underestimate on CC \*/
: > 
: > I have no idea what CC is
: 
: Me too\.  And Larry did not explain when I asked\.

I do not explain a lot of things when asked\, especially when I do not
notice that I was asked\.  :\-\)

: I assume that it is
: some indication of byte vs char\, but i could not guess \*which\* one\.

That was just a note to myself\.  Others weren't expected to understand\.  :\-\)

I believe I was using CC to mean C character\, and UC to mean utf8
character\.  The whole regex engine had grown up with the assumption
that a byte equals a character\, so when I first did the Unicode
conversion\, I had to figure out which offsets meant bytes and which
offsets meant characters\.  I sprinkled all these CC and UC comments
around just to try to keep things straight in my own head\.  Some of
those survived into the current code\.

If I recall correctly\, what the note above is basically saying is that\,
even though check\_offset\_min is calculated in terms of character
positions\, it's okay \(though not optimal\) to use it as a byte offset in
this case because you'll still get to the right character position
eventually\, because the regex in question can't match at the wrong
offset\, and as long as we don't overshoot the correct position\, we'll
still match when we get there\.

Larry

**Comment by p5pRT** [comment]
### From p5p@spam.wizbit.be

It appears that this is fixed in perl\-5\.6\.1\.

perl\-5\.6\.0 \-w rt\-3210\.pl
Name "main&#8203;::match" used only once&#8203;: possible typo at rt\-3210\.pl line 2\.
Segmentation fault

perl\-5\.6\.1 \-w rt\-3210\.pl
Name "main&#8203;::match" used only once&#8203;: possible typo at rt\-3210\.pl line 2\.

**Comment by p5pRT** [comment]
p5p@spam.wizbit.be - Status changed from 'open' to 'resolved'
